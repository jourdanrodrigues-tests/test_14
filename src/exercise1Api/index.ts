import axios from 'axios';
import currency from 'currency.js';

import type { Column } from '@/components/Table.tsx';
import { isOffline, isServerError } from '@/utils/axios.ts';

export type ServerEntry = [string, string, string];

type EntryResponse = [ServerEntry, ...ServerEntry[]];

export type Entry = {
  id: string;
  fruit: string;
  quantity: number;
  price: number;
};

const client = axios.create({
  baseURL: import.meta.env.VITE_EXERCISE_1_API_URL,
  headers: { 'Content-Type': 'application/json' },
});

const api = {
  client,
  getItems() {
    const _request = () => client.get<EntryResponse>('/list');

    return withRetryOrRethrow(_request).then(({ data }) => {
      const entries = data.slice(1).map(
        ([fruit, quantity, price]): Entry => ({
          id: fruit,
          fruit,
          quantity: +quantity,
          price: currency(price).value,
        })
      );
      const columns: Column<Entry>[] = [
        { label: 'Fruit', sortBy: 'fruit', source: 'fruit' },
        { label: 'Quantity', sortBy: 'quantity', source: 'quantity' },
        { label: 'Price', sortBy: 'price', source: 'price' },
      ];

      return { entries, columns };
    });
  },
  getHighlightIndexes() {
    const _request = () =>
      client.get<[number, number]>('/highlights').then(({ data }) => data);
    return withRetryOrRethrow(_request);
  },
};

export default api;

function withRetryOrRethrow<T = unknown>(action: () => Promise<T>): Promise<T> {
  return action().catch((reason: Error) => {
    if (!isOffline(reason) && !isServerError(reason)) throw reason;
    return withRetryOrRethrow(action);
  });
}
