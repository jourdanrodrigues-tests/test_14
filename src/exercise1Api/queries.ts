import { useQuery } from '@tanstack/react-query';
import { toast } from 'react-toastify';

import api from '@/exercise1Api';

const QueryKey = {
  entries: () => ['entries'],
  highlightIndexes: () => ['highlightIds'],
};

export function useEntriesQuery() {
  return useQuery({
    queryKey: QueryKey.entries(),
    queryFn: () => {
      toast('Fetching entries', { type: 'info' });
      return api.getItems();
    },
  });
}

export function useHighlightIndexesQuery() {
  const { data } = useQuery({
    queryKey: QueryKey.highlightIndexes(),
    queryFn: () => api.getHighlightIndexes().then((data) => new Set(data)),
    refetchInterval: 5000,
  });
  return data || new Set();
}
