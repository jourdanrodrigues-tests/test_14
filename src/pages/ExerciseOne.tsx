import Page from '@/components/Page.tsx';
import Table from '@/components/Table.tsx';
import {
  useEntriesQuery,
  useHighlightIndexes,
} from '@/exercise1Api/queries.ts';
import { useNavigator } from '@/routes.ts';

export default function ExerciseOne() {
  const navigator = useNavigator();
  const { data: { entries, columns } = { entries: [], columns: [] } } =
    useEntriesQuery();
  const highlightIndexes = useHighlightIndexes();

  return (
    <Page title="Exercise #1" className="gap-10 mt-20">
      <div className="flex flex-col gap-6">
        <button onClick={navigator.home}>Go back</button>
        <Table
          items={entries}
          columns={columns}
          rowClassName={(_, index) =>
            highlightIndexes.has(index)
              ? 'dark:bg-yellow-800 bg-yellow-100'
              : ''
          }
        />
      </div>
    </Page>
  );
}
