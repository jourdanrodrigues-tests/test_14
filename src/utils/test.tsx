import React from 'react';

import type { UseQueryResult } from '@tanstack/react-query';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render as rtlRender, waitFor } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import { expect } from 'bun:test';
import pretty from 'pretty';

type TQueryHooks<
  TData = unknown,
  TError = unknown,
  QueryResult = UseQueryResult<TData, TError>,
> = Array<{
  useHook(): QueryResult;
  validate(query: QueryResult): boolean;
}>;

type RenderOptions = {
  queryHooks?: TQueryHooks;
};

export const testQueryClient = new QueryClient();
const queryClientWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

export async function render(
  node: React.ReactElement,
  options?: RenderOptions
) {
  const hookPromises = (options?.queryHooks || []).map(
    ({ useHook, validate }) => {
      const { result } = renderHook(useHook, { wrapper: queryClientWrapper });
      return waitFor(() => validate(result.current));
    }
  );
  if (hookPromises.length > 0) {
    await Promise.all(hookPromises);
  }

  const result = rtlRender(node, {
    wrapper: ({ children }) => wrapNode(children),
  });
  return result.container;
}

function wrapNode(node: React.ReactNode): React.ReactNode {
  return queryClientWrapper({ children: node });
}

export function expectSnapshot(element: HTMLElement | null) {
  // TODO: Use @testing-library/jest-dom instead of @happy-dom
  //  The JestDOM does some HTML pretty-printing that is better for diffs
  return expect(element && pretty(element.innerHTML));
}
