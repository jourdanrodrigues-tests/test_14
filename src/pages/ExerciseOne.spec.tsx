import { waitFor } from '@testing-library/react';
import MockAdapter from 'axios-mock-adapter';
import { afterEach, beforeEach, describe, it } from 'bun:test';

import exercise1Api from '@/exercise1Api';
import ExerciseOne from '@/pages/ExerciseOne.tsx';
import { expectSnapshot, render } from '@/utils/test.tsx';

const items = [
  { id: 'Apple', fruit: 'Apple', quantity: 10, price: 1.3 },
  { id: 'Banana', fruit: 'Banana', quantity: 20, price: 2.85 },
  { id: 'Cherry', fruit: 'Cherry', quantity: 30, price: 3.12 },
  { id: 'Strawberry', fruit: 'Strawberry', quantity: 40, price: 4.83 },
  { id: 'Blueberry', fruit: 'Blueberry', quantity: 50, price: 5.01 },
  { id: 'Pineapple', fruit: 'Pineapple', quantity: 60, price: 6.45 },
  { id: 'Mango', fruit: 'Mango', quantity: 70, price: 7.62 },
  { id: 'Peach', fruit: 'Peach', quantity: 80, price: 8.15 },
  { id: 'Pear', fruit: 'Pear', quantity: 90, price: 9.81 },
  { id: 'Plum', fruit: 'Plum', quantity: 100, price: 10 },
];

describe('ExerciseOne Component', () => {
  let axiosMock: MockAdapter;

  beforeEach(() => {
    axiosMock = new MockAdapter(exercise1Api.client);
  });

  afterEach(() => {
    axiosMock.restore();
  });

  it('should match the snapshot', async () => {
    axiosMock
      .onGet('/list')
      .reply(200, [
        ['Fruit', 'Quantity', 'Price'],
        ...items.map(({ fruit, quantity, price }) => [
          fruit,
          quantity,
          `$${price}`,
        ]),
      ])
      .onGet('/highlights')
      .reply(200, [5, 3, 9, 7]);

    const result = await render(<ExerciseOne />, {
      route: ['/exercise1', '/exercise1'],
    });

    await waitFor(() => result.getByText(items[0].fruit)); // Table must be there

    expectSnapshot(result.container).toMatchSnapshot();
  });
});
