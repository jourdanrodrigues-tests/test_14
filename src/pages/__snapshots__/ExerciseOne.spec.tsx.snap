// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseOne Component should match the snapshot 1`] = `
"<div class="max-h-screen w-screen flex flex-col items-center justify-center gap-10 mt-20">
  <h1 class="block antialiased tracking-normal font-sans text-5xl font-semibold leading-tight text-inherit">Exercise #1</h1>
  <div class="flex flex-col gap-6"><button>Go back</button>
    <table class="table-auto relative w-full border">
      <thead>
        <tr class="bg-blue-gray-50 dark:bg-blue-gray-800 sticky top-0">
          <th class="px-6 py-3">
            <div class="flex gap-4">
              <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">Fruit</p>
              <div class="flex flex-col">
                <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▲</p>
                <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▼</p>
              </div>
            </div>
          </th>
          <th class="px-6 py-3">
            <div class="flex gap-4">
              <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">Quantity</p>
              <div class="flex flex-col">
                <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▲</p>
                <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▼</p>
              </div>
            </div>
          </th>
          <th class="px-6 py-3">
            <div class="flex gap-4">
              <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">Price</p>
              <div class="flex flex-col">
                <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▲</p>
                <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▼</p>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody class="divide-y">
        <tr class="even:bg-blue-gray-50/50">
          <td class="py-3 px-4">Apple</td>
          <td class="py-3 px-4">10</td>
          <td class="py-3 px-4">1.3</td>
        </tr>
        <tr class="even:bg-blue-gray-50/50">
          <td class="py-3 px-4">Banana</td>
          <td class="py-3 px-4">20</td>
          <td class="py-3 px-4">2.85</td>
        </tr>
        <tr class="even:bg-blue-gray-50/50">
          <td class="py-3 px-4">Cherry</td>
          <td class="py-3 px-4">30</td>
          <td class="py-3 px-4">3.12</td>
        </tr>
        <tr class="dark:bg-yellow-800 bg-yellow-100">
          <td class="py-3 px-4">Strawberry</td>
          <td class="py-3 px-4">40</td>
          <td class="py-3 px-4">4.83</td>
        </tr>
        <tr class="even:bg-blue-gray-50/50">
          <td class="py-3 px-4">Blueberry</td>
          <td class="py-3 px-4">50</td>
          <td class="py-3 px-4">5.01</td>
        </tr>
        <tr class="dark:bg-yellow-800 bg-yellow-100">
          <td class="py-3 px-4">Pineapple</td>
          <td class="py-3 px-4">60</td>
          <td class="py-3 px-4">6.45</td>
        </tr>
        <tr class="even:bg-blue-gray-50/50">
          <td class="py-3 px-4">Mango</td>
          <td class="py-3 px-4">70</td>
          <td class="py-3 px-4">7.62</td>
        </tr>
        <tr class="dark:bg-yellow-800 bg-yellow-100">
          <td class="py-3 px-4">Peach</td>
          <td class="py-3 px-4">80</td>
          <td class="py-3 px-4">8.15</td>
        </tr>
        <tr class="even:bg-blue-gray-50/50">
          <td class="py-3 px-4">Pear</td>
          <td class="py-3 px-4">90</td>
          <td class="py-3 px-4">9.81</td>
        </tr>
        <tr class="dark:bg-yellow-800 bg-yellow-100">
          <td class="py-3 px-4">Plum</td>
          <td class="py-3 px-4">100</td>
          <td class="py-3 px-4">10</td>
        </tr>
      </tbody>
    </table>
  </div>
</div>"
`;
